src/qilib/utils/serialization.py

```diff
@@ -18,61 +18,184 @@
+from typing import Any, Dict, Callable, ClassVar
+from qilib.data_set import MongoDataSetIO
 import numpy as np
-import base64
-import pickle
-import serialize as serialize_module
 
+TransformFunction = Callable[[Any], Any]
 
-def __object_to_base64_encoded_pickle(u):
-    b = pickle.dumps(u)
-    return base64.b64encode(b).decode('ascii')
 
+class JsonSerializeKey:
+    """ The custum value types for the JSON serializer """
+    OBJECT = '__object__'
+    CONTENT = '__content__'
 
-def __base64_encoded_pickle_to_object(c):
-    b = base64.b64decode(c.encode('ascii'))
-    return pickle.loads(b)
 
+class Encoder(JSONEncoder):
+    """ A JSON encoder """
 
-serialize_module.register_class(np.ndarray, __object_to_base64_encoded_pickle, __base64_encoded_pickle_to_object)
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
 
+        self.encoders: ClassVar[Dict[type, TransformFunction]] = {}
 
-def _bytes_to_base64(b):
-    return base64.b64encode(b).decode('ascii')
+    def default(self, o):
+        if type(o) in self.encoders:
+            return self.encoders[type(o)](o)
 
+        return JSONEncoder.default(self, o)
 
-def _base64_to_bytes(c):
-    return base64.b64decode(c.encode('ascii'))
 
+class Decoder(JSONDecoder):
+    """ A JSON dncoder """
 
-serialize_module.register_class(bytes, _bytes_to_base64, _base64_to_bytes)
+    def __init__(self):
+        super().__init__(object_hook=self._object_hook)
 
-__msgfmt = 'json'
+        self.decoders: ClassVar[Dict[type, TransformFunction]] = {}
 
+    def _object_hook(self, obj):
+        if isinstance(obj, dict):
+            if JsonSerializeKey.OBJECT in obj:
+                if obj[JsonSerializeKey.OBJECT] in self.decoders:
+                    return self.decoders[obj[JsonSerializeKey.OBJECT]](obj)
+                else:
+                    raise ValueError()
 
-def serialize(data):
-    """ Serialize a Python object to JSON
+        return obj
 
-    Special objects might be serialized to a Python string.
+
+def _encode_bytes(data: Any) -> Dict[str, Any]:
+    return {JsonSerializeKey.OBJECT: bytes.__name__, JsonSerializeKey.CONTENT: data.decode('utf-8')}
+
+
+def _decode_bytes(data: Dict[str, Any]) -> bytes:
+    return data[JsonSerializeKey.CONTENT].encode('utf-8')
+
+
+class Serializer:
+    def __init__(self, encoders=None, decoders=None):
+        self.encoder = Encoder()
+        self.decoder = Decoder()
+
+        if encoders is None:
+            encoders = {}
+        self.encoder.encoders = encoders
+        if decoders is None:
+            decoders = {}
+        self.decoder.decoders = decoders
+
+        self.register_encoder(bytes, _encode_bytes)
+        self.register_decoder(bytes.__name__, _decode_bytes)
+        self.register_encoder(np.ndarray, MongoDataSetIO.encode_numpy_array)
+        self.register_decoder(np.array.__name__, MongoDataSetIO.decode_numpy_array)
+
+    def register_encoder(self, type_: type, encode_func: TransformFunction):
```

Comment:
Make `type_` consistent in naming with `type_name` from `register_decoder`. Also: why not a single method
```
def register(self, encode_function, decode_function, type, type_name) ?
```

Reply:
`type_` and `type_name` are two different types, a type object and a string object, and therefore they aren't named in the same way. Furthermore, `type` is a Python keyword, so it is prohibited to use that as an identifier/name.

I thought about having one method for registering both an encoder and decoder. But I wasn't sure if a type should always have a decoder. But having them separate leaves this option up to the user. What do you think?

Reply:
Anyways I have made a `register` method to do register both an encoder and decoder as that does indeed make more sense if we're writing a serializer.